ibis/common/egraph.py

```diff
@@ -0,0 +1,582 @@
+
+    def add(self, node: Node) -> ENode:
+        """Add a node to the egraph.
+
+        The node is converted to an enode and added to the egraph. If the enode is
+        already present in the egraph, then the canonical enode is returned.
+
+        Parameters
+        ----------
+        node :
+            The node to add to the egraph.
+
+        Returns
+        -------
+        enode :
+            The canonical enode.
+        """
+        enode = self._as_enode(node)
+        if enode in self._eclasses:
+            return self._eclasses.find(enode)
+
+        args = []
+        for arg in enode.args:
+            if isinstance(arg, ENode):
+                args.append(self.add(arg))
+            else:
+                args.append(arg)
+
+        enode = ENode(enode.head, args)
+        self._eclasses.add(enode)
+        self._etables[enode.head][enode] = tuple(args)
+
+        return enode
+
+    def union(self, node1: Node, node2: Node) -> ENode:
+        """Union two nodes in the egraph.
+
+        The nodes are converted to enodes which must be present in the egraph.
+        The eclasses of the nodes are merged and the canonical enode is returned.
+
+        Parameters
+        ----------
+        node1 :
+            The first node to union.
+        node2 :
+            The second node to union.
+
+        Returns
+        -------
+        enode :
+            The canonical enode.
+        """
+        enode1 = self._as_enode(node1)
+        enode2 = self._as_enode(node2)
+        return self._eclasses.union(enode1, enode2)
+
+    def _match_args(self, args, patargs):
+        """Match the arguments of an enode against a pattern's arguments.
+
+        An enode matches a pattern if each of the arguments are:
+        - both leaf values and equal
+        - both enodes and in the same eclass
+        - an enode and a variable, in which case the variable gets bound to the enode
+
+        Parameters
+        ----------
+        args : tuple
+            The arguments of the enode. Since an enode is a ground term, the arguments
+            are either enodes or leaf values.
+        patargs : tuple
+            The arguments of the pattern. Since a pattern is a flat term (flattened
+            using auxilliary variables), the arguments are either variables or leaf
+            values.
+
+        Returns
+        -------
+        dict[str, Any] :
+            The mapping of variable names to enodes or leaf values.
+        """
+        subst = {}
+        for arg, patarg in zip(args, patargs):
+            if isinstance(patarg, Variable):
+                if patarg.name is None:
+                    pass
+                elif isinstance(arg, ENode):
+                    subst[patarg.name] = self._eclasses.find(arg)
+                else:
+                    subst[patarg.name] = arg
+            elif isinstance(arg, ENode):
+                if self._eclasses.find(arg) != self._eclasses.find(arg):
+                    return None
```

Comment:
There are two uncovered branches in `_match_args` here. Should we add tests for these cases?

Reply:
Perhaps they are not needed, going to double check and cover if required.

Reply:
The first uncovered branch is not possible since variables must be named, I removed that.

The second branch is *currently* not possible due to the pattern flattening the `rhs` is either a Variable or a leaf value. 
BUT I commented out that branch instead of removing because we may choose to allow ENodes in patterns in the future depending on what kind of patterns do we want to allow and how do we convert that to queries.